server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java

```diff
@@ -40,26 +41,64 @@
 
     @Override
     public byte readByte() throws IOException {
-        if (offset() >= bytesReference.length()) {
-            throw new EOFException();
-        }
         maybeNextSlice();
         return slice.bytes[slice.offset + (sliceIndex++)];
     }
 
+    @Override
+    public short readShort() throws IOException {
+        if (slice.length - sliceIndex >= 2) {
+            sliceIndex += 2;
+            return Numbers.bytesToShort(slice.bytes, slice.offset + sliceIndex - 2);
+        } else {
+            // slow path
+            return super.readShort();
+        }
+    }
+
+    @Override
+    public int readInt() throws IOException {
+        if (slice.length - sliceIndex >= 4) {
+            sliceIndex += 4;
+            return Numbers.bytesToInt(slice.bytes, slice.offset + sliceIndex - 4);
+        } else {
+            // slow path
+            return super.readInt();
+        }
+    }
+
+    @Override
+    public long readLong() throws IOException {
+        if (slice.length - sliceIndex >= 8) {
+            sliceIndex += 8;
+            return Numbers.bytesToLong(slice.bytes, slice.offset + sliceIndex - 8);
+        } else {
+            // slow path
+            return super.readLong();
+        }
+    }
+
     protected int offset() {
         return sliceStartOffset + sliceIndex;
     }
 
     private void maybeNextSlice() throws IOException {
-        while (sliceIndex == slice.length) {
-            sliceStartOffset += sliceIndex;
+        if (sliceIndex == slice.length) {
```

Comment:
Its worth leaving a comment here that we're being super intentional about moving the slow stuff to another method so the `if` statement is inlined.

Reply:
++ added one.